### PR TITLE
Merge Nullability Analysis; Changes according to #7546

### PR DESCRIPTION
**Pull request by noti0na1**
## Changes according to #7546
* Add extractor for JavaNull ops
* Remove useless imports
* Reduce side effects
* Update types in DottyPredef
* Fix typo

## Merge Nullability Analysis
* Merge Nullability Analysis
* Disabled one test in `nullable.scala`; it requires implementing `notNull`

## Others
* Update `Types.widenUnion`

### CODE REVIEW

**Comment by noti0na1** [comment]
I rewrote the `widenUnion` as below:

```scala
case tp @ OrType(lhs, rhs) =>
  tp match {
    case OrNull(tp1) =>
      // Don't widen `T|Null`, since otherwise we wouldn't be able to infer nullable unions.
      val tp1Widen = tp1.widenUnion
      if (tp1Widen.isRef(defn.AnyClass)) tp1Widen
      else tp.derivedOrType(tp1Widen, defn.NullType)
    case _ =>
      ctx.typeComparer.lub(lhs.widenUnion, rhs.widenUnion, canConstrain = true) match {
        case union: OrType => union.join
        case res => res
      }
  }
```

Because the original implementation will infer incorrect type for this case:

```scala
class A
class B

val x: (A | Null) | B = ???
val y = x // the lub for A and B is AnyRef, but Any | Null is inferrred 
```  

Also, I don't think using `stripNull` is very expensive here, and this is more readable.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -1610,12 +1610,18 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] w
...
             // the Java argument has type 'Any', and the Scala argument has type 'Object' or
             // 'Object|Null', depending on whether explicit nulls are enabled.
-            lazy val formal2IsObject =
-              if (ctx.explicitNulls) formal2.isNullableUnion && formal2.stripNull(ctx).isRef(ObjectClass)
-              else formal2.isRef(ObjectClass)
-            lazy val formal1IsObject =
-              if (ctx.explicitNulls) formal1.isNullableUnion && formal1.stripNull(ctx).isRef(ObjectClass)
+            def formal1IsObject =
+              if (ctx.explicitNulls) formal1 match {
+                case OrNull(formal1b) => formal1b.isRef(ObjectClass)
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1615]: nice

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 2]
```diff
@@ -1072,44 +1105,18 @@ object Types {
...
-          def isNullableUnionFast(tp: Type): Boolean = tp match {
-            case orTpe: OrType if orTpe.tp2.isNullType => true
-            case _ => false
-          }
-
-          if (ctx.explicitNulls) {
-            // Don't widen `T|Null`, since otherwise we wouldn't be able to infer nullable unions.
-            // This part relies on the postcondition of widenUnion: the result is either a
-            // non-union type, or a nullable union type where the rhs is `Null` type.
-            if (rhs.isNullType) ensureNullableUnion(lhs.widenUnion, tp)
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala:1099]: Did we agree that we don't need the more efficient code? Why was that?

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 3]
```diff
@@ -1437,6 +1444,11 @@ object Types {
       case _ => true
     }
 
+    /** Is this (an alias of) the `scala.Null` type? */
+    final def isNull(given Context) =
+      isRef(defn.NullClass)
+      || classSymbol.name == tpnme.Null // !!! temporary kludge for being able to test without the explicit nulls PR
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: Do we still need this second condition? The comment seems to indicate we no longer do.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 4]
```diff
@@ -2924,6 +2936,28 @@ object Types {
       else apply(tp1, tp2)
   }
 
+  object OrNull {
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala:2944]: Could you please add a comment saying what this is?

e.g.
```Scala
An extractor object to pattern match against a nullable union.
e.g.

(tp: Type) match {
  case OrNull(tp1) => // tp had the form `tp1 | Null`
  case _ => // tp was not a nullable union
}
```

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 5]
```diff
@@ -2924,6 +2936,28 @@ object Types {
...
+      OrType(tp, defn.NullType)
+    def unapply(tp: Type)(given ctx: Context): Option[Type] =
+    if (ctx.explicitNulls) {
+      val tp1 = tp.stripNull
+      if tp1 ne tp then Some(tp1) else None
+    }
+    else None
+  }
+
+  object OrJavaNull {
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala:2962]: Ditto, we need a comment here.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 6]
```diff
@@ -2924,6 +2936,28 @@ object Types {
       else apply(tp1, tp2)
   }
 
+  object OrNull {
+    def apply(tp: Type)(given Context) =
+      OrType(tp, defn.NullType)
+    def unapply(tp: Type)(given ctx: Context): Option[Type] =
+    if (ctx.explicitNulls) {
+      val tp1 = tp.stripNull
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala:2949]: nice

**Review by abeln — APPROVED** [review]
Thanks! Feel free to merge after addressing the comments below.

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 3 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Types.scala]: Removed

**Comment by noti0na1** [comment]
@abeln 

```
-- [E007] Type Mismatch Error: tests/explicit-nulls/pos/ref-eq.scala:9:8 -----------------------------------------------
9 |  x1.ne(1)
  |        ^
  |        Found:    Int(1)
  |        Required: Object
-- [E007] Type Mismatch Error: tests/explicit-nulls/pos/ref-eq.scala:10:8 ----------------------------------------------
10 |  x1.eq(null)
   |        ^^^^
   |        Found:    Null
   |        Required: Object
-- [E007] Type Mismatch Error: tests/explicit-nulls/pos/ref-eq.scala:11:8 ----------------------------------------------
11 |  x1.ne(null)
   |        ^^^^
   |        Found:    Null
   |        Required: Object
-- [E007] Type Mismatch Error: tests/explicit-nulls/pos/ref-eq.scala:12:8 ----------------------------------------------
12 |  x1.eq(1) // ok: implicit conversion from int to Integer
   |        ^
   |        Found:    Int(1)
   |        Required: Object
-- [E007] Type Mismatch Error: tests/explicit-nulls/pos/ref-eq.scala:13:8 ----------------------------------------------
13 |  x1.ne(1) // ok: ditto
   |        ^
   |        Found:    Int(1)
   |        Required: Object
-- [E007] Type Mismatch Error: tests/explicit-nulls/pos/ref-eq.scala:14:8 ----------------------------------------------
14 |  x1.eq(x2)
   |        ^^
   |        Found:    (String | Null)(Test.this.x2)
   |        Required: Object
-- [E007] Type Mismatch Error: tests/explicit-nulls/pos/ref-eq.scala:15:8 ----------------------------------------------
15 |  x1.ne(x2)
   |        ^^
   |        Found:    (String | Null)(Test.this.x2)
   |        Required: Object
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:18:7 -------------------------------------------
18 |  null.eq("hello")
   |  ^^^^^^^
   |  value eq is not a member of Null
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:19:7 -------------------------------------------
19 |  null.ne("hello")
   |  ^^^^^^^
   |  value ne is not a member of Null
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:20:7 -------------------------------------------
20 |  null.eq(null)
   |  ^^^^^^^
   |  value eq is not a member of Null
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:21:7 -------------------------------------------
21 |  null.ne(null)
   |  ^^^^^^^
   |  value ne is not a member of Null
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:22:7 -------------------------------------------
22 |  null.eq(x2)
   |  ^^^^^^^
   |  value eq is not a member of Null
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:23:7 -------------------------------------------
23 |  null.ne(x2)
   |  ^^^^^^^
   |  value ne is not a member of Null
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:25:5 -------------------------------------------
25 |  x2.eq(null)
   |  ^^^^^
   |  value eq is not a member of String | Null
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:26:5 -------------------------------------------
26 |  x2.ne(null)
   |  ^^^^^
   |  value ne is not a member of String | Null
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:27:5 -------------------------------------------
27 |  x2.eq(x1)
   |  ^^^^^
   |  value eq is not a member of String | Null
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:28:5 -------------------------------------------
28 |  x2.ne(x1)
   |  ^^^^^
   |  value ne is not a member of String | Null
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:29:5 -------------------------------------------
29 |  x2.eq(x2)
   |  ^^^^^
   |  value eq is not a member of String | Null
-- [E008] Member Not Found Error: tests/explicit-nulls/pos/ref-eq.scala:30:5 -------------------------------------------
30 |  x2.ne(x2)
   |  ^^^^^
   |  value ne is not a member of String | Null
Compilation failed for: 'tests/explicit-nulls/pos/ref-eq.scala'          
```

```
-- [E008] Member Not Found Error: tests/explicit-nulls/run/subtype-any.scala:5:16 --------------------------------------
5 |    assert(null.eq(null))
  |           ^^^^^^^
  |           value eq is not a member of Null
-- [E008] Member Not Found Error: tests/explicit-nulls/run/subtype-any.scala:6:17 --------------------------------------
6 |    assert(!null.ne(null))
  |            ^^^^^^^
  |            value ne is not a member of Null
-- [E008] Member Not Found Error: tests/explicit-nulls/run/subtype-any.scala:8:17 --------------------------------------
8 |    assert(!null.eq("hello"))
  |            ^^^^^^^
  |            value eq is not a member of Null
-- [E008] Member Not Found Error: tests/explicit-nulls/run/subtype-any.scala:9:16 --------------------------------------
9 |    assert(null.ne("hello"))
  |           ^^^^^^^
  |           value ne is not a member of Null
-- [E008] Member Not Found Error: tests/explicit-nulls/run/subtype-any.scala:11:17 -------------------------------------
11 |    assert(!null.eq(4))
   |            ^^^^^^^
   |            value eq is not a member of Null
-- [E008] Member Not Found Error: tests/explicit-nulls/run/subtype-any.scala:12:16 -------------------------------------
12 |    assert(null.ne(4))
   |           ^^^^^^^
   |           value ne is not a member of Null
-- [E007] Type Mismatch Error: tests/explicit-nulls/run/subtype-any.scala:14:23 ----------------------------------------
14 |    assert(!"hello".eq(null))
   |                       ^^^^
   |                       Found:    Null
   |                       Required: Object
-- [E007] Type Mismatch Error: tests/explicit-nulls/run/subtype-any.scala:15:22 ----------------------------------------
15 |    assert("hello".ne(null))
   |                      ^^^^
   |                      Found:    Null
   |                      Required: Object
-- [E007] Type Mismatch Error: tests/explicit-nulls/run/subtype-any.scala:17:17 ----------------------------------------
17 |    assert(!4.eq(null))
   |                 ^^^^
   |                 Found:    Null
   |                 Required: Object
-- [E007] Type Mismatch Error: tests/explicit-nulls/run/subtype-any.scala:18:16 ----------------------------------------
18 |    assert(4.ne(null))
   |                ^^^^
   |                Found:    Null
   |                Required: Object
-- [E008] Member Not Found Error: tests/explicit-nulls/run/subtype-any.scala:21:13 -------------------------------------
21 |    assert(x.eq(null))
   |           ^^^^
   |           value eq is not a member of String | Null
-- [E008] Member Not Found Error: tests/explicit-nulls/run/subtype-any.scala:22:14 -------------------------------------
22 |    assert(!x.ne(null))
   |            ^^^^
   |            value ne is not a member of String | Null
-- [E008] Member Not Found Error: tests/explicit-nulls/run/subtype-any.scala:25:15 -------------------------------------
25 |    assert(!x2.eq(null))
   |            ^^^^^
   |            value eq is not a member of AnyRef | Null
-- [E008] Member Not Found Error: tests/explicit-nulls/run/subtype-any.scala:26:14 -------------------------------------
26 |    assert(x2.ne(null))
   |           ^^^^^
   |           value ne is not a member of AnyRef | Null
Compilation failed for: 'tests/explicit-nulls/run/subtype-any.scala' 
```

**Pull request merged**